source/common/event/scaled_timer.cc

```diff
@@ -0,0 +1,207 @@
+#include "envoy/event/dispatcher.h"
+#include "envoy/event/timer.h"
+#include "common/event/scaled_timer.h"
+#include <algorithm>
+#include <chrono>
+
+namespace Envoy {
+namespace Event {
+
+class ScaledTimerManager::ScaledTimerImpl final : public Timer {
+public:
+  struct ActiveTimer {
+    const MonotonicTime start;
+    const std::chrono::microseconds delay;
+    const ScopeTrackedObject* object;
+  };
+
+  ScaledTimerImpl(ScaledTimerManager& manager, TimerCb callback);
+
+  ~ScaledTimerImpl() override;
+
+  /**
+   * Run the timer callback on the given dispatcher.
+   */
+  void dispatch(Dispatcher& dispatcher);
+  const absl::optional<ActiveTimer>& active() const;
+
+  // Timer impl:
+  void enableTimer(const std::chrono::milliseconds& delay,
+                   const ScopeTrackedObject* object) override;
```

Comment:
We talked about the issue of this timer scaling down to 0, not down to a specified minimum.  I'm not sure that the Timer interface applies for scaled timers.

Reply:
Yep, I think this PR is going to be superseded by #11806. See the ScaledTimer interface [here](https://github.com/envoyproxy/envoy/pull/11806/files#diff-938f3cc745b8fe3c5362ed015d6574ffR63)